src/main/java/org/killbill/billing/plugin/payment/PaymentTestPluginApi.java

```diff
@@ -0,0 +1,384 @@
+    int getSleepValue(final String methodCalled,
+                      final Iterable<PluginProperty> pluginProperties) {
+        int sleep = 0;
+        if (pluginProperties != null) {
+            // look for sleep in 'one time config'
+            final Optional<PluginProperty> actionFromProperties =
+                    StreamSupport.stream(pluginProperties.spliterator(), false)
+                                 // find a plugin property
+                                 .filter(p -> {
+                                     try {
+                                         // where key is a known action
+                                         TestingStates.Actions a = TestingStates.Actions.valueOf(p.getKey());
+                                         // and value match all or the method called
+                                         if (a.compareTo(TestingStates.Actions.ACTION_SLEEP) == 0 && (
+                                                 p.getValue() == null || ((String) p.getValue()).length() == 0 ||
+                                                         ((String) p.getValue()).compareTo("*") == 0 ||
+                                                         ((String) p.getValue()).compareTo(methodCalled) == 0)) {
+                                             return true;
+                                         }
+                                     }
+                                     catch (final IllegalArgumentException ignore) {}
+                                     return false;
+                                 })
+                                 .findFirst();
+            if (actionFromProperties.isPresent()) {
+                // find sleep value
+                sleep = StreamSupport.stream(pluginProperties.spliterator(), false)
+                                     .filter(property -> TestingStates.SLEEP_PLUGIN_CONFIG_PARAM.compareTo(property.getKey()) == 0)
+                                     .findFirst().map(property -> Integer.valueOf((String) property.getValue()))
+                                     .orElse(0);
+            }
+        }
+        if (sleep == 0) {
+            // look for sleep in global config
+            final Integer globalSleep = (this.testingStates.getSleeps().get("*") != null)
+                    ? this.testingStates.getSleeps().get("*") : this.testingStates.getSleeps().get(methodCalled);
+            if (globalSleep != null && globalSleep.compareTo(this.noSleep) > 0) {
+                sleep = globalSleep;
+            }
+        }
+        return sleep;
+    }
+
+    private PaymentPluginStatus handleState(final Iterable<PluginProperty> pluginProperties) throws PaymentPluginApiException {
+        final String methodCalled = Thread.currentThread().getStackTrace()[2].getMethodName();
+
+        final TestingStates.Actions action = getAction(methodCalled, pluginProperties);
+        final int sleep = getSleepValue(methodCalled, pluginProperties);
+        if (sleep > 0) {
```

Comment:
The main comment here is that your enum `TestingStates.Actions` contains both the various action and the `ACTION_SLEEP` (which is handled differently). Perhaps, it would make sense to remove it from the enum and keep the separate handling the way you have it?

Reply:
as you want, right now it's convenient to validated the value of request action, for example here:
https://github.com/laurentbh/killbill-payment-test-plugin/blob/initial-release/src/main/java/org/killbill/billing/plugin/payment/resources/PaymentTestResource.java#L40